### DISCUSSION TITLE
Javascript (server side) bindings via n-api

### DISCUSSION THREAD

**Discussion by universalmind303** [delta-io/delta-rs#3248] [General]
# Description

I'd like a similar sdk as the python sdk but available in server side js (node, deno, bun, ...).  The sdk should closely resemble the python one, and only deviate when either necessary due to napi limitations, or when it is unidiomatic in JS.

[napi.rs](https://napi.rs/) is very similar in nature to pyo3 and is relatively easy to use. 

**Use Case**
I want to read delta files via a js application.

**Related Issue(s)**

https://github.com/pola-rs/nodejs-polars/issues/176

**Comment by universalmind303** [comment]
@rtyler I'd be happy to contribute to this one. Feel free to assign it to me.

**Comment by ion-elgreco** [comment]
@universalmind303 I tried creating a reader with polars rust directly but the Scan is not flexible enough to read schema evolved table unfortunately : (

So curious to see how you are planning to do this with polars 😄

**Comment by universalmind303** [comment]
> @universalmind303 I tried creating a reader with polars rust directly but the Scan is not flexible enough to read schema evolved table unfortunately : (
> 
> So curious to see how you are planning to do this with polars 😄

for polars, there is a way to push down some operations to the datasource during planning. `AnonymousScan`. It doesn't support all of the pushdowns, but it can likely be improved to support all of the same pushdowns as parquet. If it's all wired up correctly, the planner should handle all of the schema evolutions for you.

**Comment by ion-elgreco** [comment]
@universalmind303 I tried using a LogicalPlan::FileScanfor parquet but polars will error out during collection, since it concats the dfs instead of doing a diagonal concat. Also if you provide a schema to the reader and a column is not in the parquet, it will error out instead of creating a null array

Here you can see old branch where I was playing around with the scan: https://github.com/ion-elgreco/polars-deltalake/blob/feat/cloud_reads/python/src/lib.rs

**Comment by TheKnightCoder** [comment]
What is the update on this? Would also love this to use  read/write delta with nodejs , huge unlock with the amount of js devs imo.

**Comment by universalmind303** [comment]
> What is the update on this? Would also love this to use read/write delta with nodejs , huge unlock with the amount of js devs imo.

I had an initial POC for it [here](https://github.com/delta-io/delta-rs/pull/2345), but since then, the python API has changed quite a bit. I made a bit of progress getting it to feature parity with the python one, but not quite there yet. 

I'm hoping to find some time to work on this again soon. If anyone else wants to work on it in the meantime, by all means!

**Comment by am2222** [comment]
may I ask if there is any plans for this?

**Reply by roeap** [reply]
There are no real concrete ones. I recently did this in another project, and the experience was quite nice! Since then, I have wanted to do this if I ever find some time. One thing that should definitely happen before is that we stabilise our APIs (much of this has happened).
